// @tsed/common/lib/mvc/models/RequestContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const RequestLogger_1 = require("./RequestLogger");
class RequestContext extends Map {
    constructor({ id, logger, url, ignoreUrlPatterns = [] }) {
        super();
        /**
         * Date when request have been handled by the server. @@RequestLogger@@ use this date to log request duration.
         */
        this.dateStart = new Date();
        /**
         * The request container used by the Ts.ED DI. It contain all services annotated with `@Scope(ProviderScope.REQUEST)`
         */
        this.container = new di_1.LocalsContainer();
        this.id = id;
        this.logger = new RequestLogger_1.RequestLogger(logger, {
            id,
            startDate: this.dateStart,
            url,
            ignoreUrlPatterns
        });
    }
    destroy() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.container.destroy();
        });
    }
}
exports.RequestContext = RequestContext;
